Java/core/src/main/java/com/amazon/randomcutforest/RandomCutForest.java

```
@@ -1004,42 +1062,107 @@ public RandomCutForest build() {
+
+                for (IComponentModel<?> model : forest.componentModels) {
+                    SamplerPlusTree<Integer> samplerPlusTree = (SamplerPlusTree<Integer>) model;
+                    CompactSamplerPlusTreeState modelState = new CompactSamplerPlusTreeState();
+                    CompactSampler sampler = (CompactSampler) samplerPlusTree.getSampler();
+                    modelState.setSamplerState(samplerMapper.toState(sampler));
+                    if (forest.saveTreeData) {
+                        CompactRandomCutTreeDouble tree = (CompactRandomCutTreeDouble) samplerPlusTree.getTree();
+                        modelState.setTreeState(treeMapper.toState(tree));
+                    }
+                    componentModelStates.add(modelState);
+                }
+                state.setComponentModelStates(componentModelStates);
+            } else {
+                List<Object> componentModelStates = new ArrayList<>();
+                IContextualStateMapper<SimpleStreamSampler<double[]>, ?, ?> mapper;
+                if (forest.storeSequenceIndexesEnabled) {
+                    mapper = new SimpleStreamSampler.ArraySequentialStateMapper();
+                } else {
+                    mapper = new SimpleStreamSampler.ArrayStateMapper();
+                }
+
+                for (IComponentModel<?> model : forest.componentModels) {
+                    SamplerPlusTree<double[]> samplerPlusTree = (SamplerPlusTree<double[]>) model;
+                    SimpleStreamSampler<double[]> sampler = (SimpleStreamSampler<double[]>) samplerPlusTree
+                            .getSampler();
+                    componentModelStates.add(mapper.toState(sampler));
+                }
+            }
+
+            return state;
+        }
+
+        @Override
+        public RandomCutForest toModel(RandomCutForestState state) {
+            return null;
```

Can we add a todo or maybe throw UnsupportedOperationException?

---

Done.